### PR TITLE
Optimize the flow followed for each API request

### PR DESCRIPTION
**Pull request by AdriiiPRodri**
|Related issue|
|---|
|#8594|

## Description

Hi team,

We have cached the `check_token` function. With this, we have been able to improve the performance of each API call.

### 4.2 (200 request -> `GET /groups/default/agents`)

old_results

### feature/8594-cache-check-token (200 request -> `GET /groups/default/agents`)

new_results

### Size (200 elements)

size

### Observations

In the case of the request (`GET /groups/default/agents`), almost all the time used by the request corresponds to the `check_token` function. We have executed three consecutive times the script that makes the request with each of the versions and we have obtained the mean time of `check_token` for each execution.

The mean of the three runs in `4.2` is: **0.039297441s**
The mean of the three runs in `feature/8594-cache-check-token` is: **0.001650225s**
Improvement: 0.039297441 / 0.001650225 = **23.81338363** -> About 23 times faster

Finally, we see that the amount of space used by 200 tokens is around 9 KB and this size increases linearly. In addition, **tokens stored in the cache will be invalidated once they expire**. **We have set a maximum of 4500 elements in the cache so the maximum cache size will be 4500 * 9 / 200 = 202.5 KB.**

Regards.

### CODE REVIEW

File: framework/wazuh/core/agent.py [hunk 1]
```diff
@@ -1250,10 +1251,11 @@ def send_restart_command(agent_id: str = '', agent_version: str = '') -> str:
 @common.context_cached('system_agents')
 def get_agents_info():
     """Get all agent IDs in the system."""
+    agent_regex = re.compile(r"^([^# ]\d+) [^!].* .* .*$", re.MULTILINE)
```
Comment (davidjiglesias) [inline framework/wazuh/core/agent.py:1254]:
```suggestion
    agent_regex = re.compile(r"^(\d{3,}) [^!].* .* .*$", re.MULTILINE)
```

File: framework/wazuh/core/agent.py [hunk 2]
```diff
@@ -1250,10 +1251,11 @@ def send_restart_command(agent_id: str = '', agent_version: str = '') -> str:
 @common.context_cached('system_agents')
 def get_agents_info():
     """Get all agent IDs in the system."""
```
Comment (davidjiglesias) [inline framework/wazuh/core/agent.py:1253]: Move regex compile outside the function so it is done only once.

**Review by davidjiglesias — COMMENTED** [review]
roles : list -> Tuple

**Review by davidjiglesias — APPROVED** [review]

**Comment by Selutario** [comment]
## Issue update

Hi team,

I did some manual testing just to double-check that it works as expected:

**Configuration and other tests:**
- [x] Run API request in master. Token is decoded only the 1st time.
- [x] Change `auth_token_exp_timeout` to 30s. Cache lasts the same that the token itself.
- [x] Tokens are always decoded when performing requests to worker's APIs.
- [x] Cache expires when RBAC mode is changed through master's API.
- [x] Cache expires when RBAC mode is changed through worker's API.
- [x] Updating one user's role does not expire the cache of other users.

**Roles:**
- [x] Token & cache expires when adding a role to the user.
- [x] Cache expires when updating a role that was assigned to the user.
- [x] Token & cache expires when deleting a role that was assigned to the user.

**Rules:**
- [x] Cache expires when adding rules to a role that is assigned to the user.
- [x] Cache expires when updating a rule that was assigned to the user's role.
- [x] Cache expires when deleting rules from a role that is assigned to the user.

**Policies:**
- [x] Token & cache expires when adding policies to a role that is assigned to the user.
- [x] Token & cache expires when deleting policies from a role that is assigned to the user.
- [x] Token & cache expires when updating a policy that was assigned to the user's role.

Regards,
Selu.

**Pull request merged**